x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -241,10 +252,58 @@ public void updateProfileData(UpdateProfileDataRequest request, ActionListener<A
             return;
         }
 
-        doUpdate(
-            buildUpdateRequest(request.getUid(), builder, request.getRefreshPolicy(), request.getIfPrimaryTerm(), request.getIfSeqNo()),
-            listener.map(updateResponse -> AcknowledgedResponse.TRUE)
-        );
+        getVersionedDocument(request.getUid(), ActionListener.wrap(doc -> {
+            validateProfileSize(doc, request, maxProfileSizeInBytes);
+
+            doUpdate(
+                buildUpdateRequest(request.getUid(), builder, request.getRefreshPolicy(), request.getIfPrimaryTerm(), request.getIfSeqNo()),
+                listener.map(updateResponse -> AcknowledgedResponse.TRUE)
+            );
+        }, listener::onFailure));
+    }
+
+    static void validateProfileSize(VersionedDocument doc, UpdateProfileDataRequest request, int limit) {
+        if (doc == null) {
+            return;
+        }
+        Map<String, Object> labels = combineMaps(doc.doc.labels(), request.getLabels());
+        Map<String, Object> data = combineMaps(mapFromBytesReference(doc.doc.applicationData()), request.getData());
+        int actualSize = serializationSize(labels) + serializationSize(data);
+        if (actualSize > limit) {
+            throw new ElasticsearchException(
```

Comment:
I think this should be an `IllegalArgumentException` and result in a 400 since it's caused by an issue triggered by a series of bad requests. WDYT?